roachtest: splits/load/uniform/nodes=3 failed

SHA: https://github.com/cockroachdb/cockroach/commits/6885730c58f9a45511f92be95e94129005d6b875

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=splits/load/uniform/nodes=3 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1099559&tab=buildLog

```
The test failed on master:
	test.go:727,cluster.go:1562,split.go:196,split.go:58: kv.kv has 1 ranges, expected between 21 and 2147483647 splits

```

This repros right away:

```
--- FAIL: splits/load/uniform/nodes=3 (96.27s)
        test.go:727,cluster.go:1562,split.go:196,split.go:58: kv.kv has 1 ranges, expected between 21 and 2147483647 splits
```

Should be easy to track down. I'll take a look.

--- a/pkg/storage/merge_queue.go
+++ b/pkg/storage/merge_queue.go
@@ -233,7 +233,6 @@ func (mq *mergeQueue) process(
 	}
 
 	lhsQPS := lhsRepl.GetSplitQPS()
-	timeSinceLastReq := lhsRepl.store.Clock().PhysicalTime().Sub(lhsRepl.GetLastRequestTime())
 	rhsDesc, rhsStats, rhsQPS, err := mq.requestRangeStats(ctx, lhsDesc.EndKey.AsRawKey())
 	if err != nil {
 		return err
@@ -250,10 +249,9 @@ func (mq *mergeQueue) process(
 	}
 	mergedStats := lhsStats
 	mergedStats.Add(rhsStats)
+
 	var mergedQPS float64
-	// Consider the mergedReqRate only if the last request is recent. If no request has been
-	// found for over 10 seconds, the merge queue should feel free to merge the ranges.
-	if lhsRepl.SplitByLoadEnabled() && timeSinceLastReq <= (10*time.Second) {
+	if lhsRepl.SplitByLoadEnabled() {
 		mergedQPS = lhsQPS + rhsQPS
 	}
 
@@ -261,9 +259,11 @@ func (mq *mergeQueue) process(
 	// Use a lower threshold for load based splitting so we don't find ourselves
 	// in a situation where we keep merging ranges that would be split soon after
 	// by a small increase in load.
-	if ok, _ := shouldSplitRange(mergedDesc, mergedStats, mergedQPS,
-		lhsRepl.SplitByLoadQPSThreshold()/2, lhsRepl.GetMaxBytes(), sysCfg); ok {
-		log.VEventf(ctx, 2, "skipping merge: merged range %s would need to be split (estimated size, estimated QPS: %d, %v)",
+	loadBasedSplitPossible := lhsRepl.SplitByLoadQPSThreshold() < 2*mergedQPS
+	if ok, _ := shouldSplitRange(mergedDesc, mergedStats, lhsRepl.GetMaxBytes(), sysCfg); ok || loadBasedSplitPossible {
+		log.VEventf(ctx, 2,
+			"skipping merge to avoid thrashing: merged range %s may split "+
+				"(estimated size, estimated QPS: %d, %v)",
 			mergedDesc, mergedStats.Total(), mergedQPS)
 		return nil
 	}

--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -18,7 +18,6 @@ import (
 	"bytes"
 	"context"
 	"fmt"
-	"math/rand"
 	"reflect"
 	"sync/atomic"
 	"time"
@@ -418,14 +417,8 @@ type Replica struct {
 		draining bool
 	}
 
-	// Split keeps information for load-based splitting.
-	splitMu struct {
-		syncutil.Mutex
-		lastReqTime time.Time // most recent time recorded by requests.
-		count       int64     // reqs since last nanos
-		qps         float64   // last reqs/s rate
-		splitFinder *split.Finder
-	}
+	// loadBasedSplitter keeps information about load-based splitting.
+	loadBasedSplitter split.Decider
 
 	unreachablesMu struct {
 		syncutil.Mutex
@@ -779,26 +772,13 @@ func (r *Replica) GetMVCCStats() enginepb.MVCCStats {
 }
 
 // GetSplitQPS returns the Replica's queries/s request rate.
-// The value returned represents the QPS recorded at the time of the
-// last request which can be found using GetLastRequestTime().
+//
 // NOTE: This should only be used for load based splitting, only
 // works when the load based splitting cluster setting is enabled.
 //
 // Use QueriesPerSecond() for current QPS stats for all other purposes.
 func (r *Replica) GetSplitQPS() float64 {
-	r.splitMu.Lock()
-	defer r.splitMu.Unlock()
-	return r.splitMu.qps
-}
-
-// GetLastRequestTime returns the most recent time in nanos
-// when the last rate was recorded.
-// NOTE: This should only be used for load based splitting, only
-// works when the load based splitting cluster setting is enabled.
-func (r *Replica) GetLastRequestTime() time.Time {
-	r.splitMu.Lock()
-	defer r.splitMu.Unlock()
-	return r.splitMu.lastReqTime
+	return r.loadBasedSplitter.LastQPS(timeutil.Now())
 }
 
 // ContainsKey returns whether this range contains the specified key.
@@ -1158,17 +1138,14 @@ func (r *Replica) beginCmds(
 
 	// Handle load-based splitting.
 	if r.SplitByLoadEnabled() {
-		r.splitMu.Lock()
-		r.splitMu.count += int64(len(ba.Requests))
-		record, split := r.needsSplitByLoadLocked()
-		if record {
-			if boundarySpan := spans.BoundarySpan(spanset.SpanGlobal); boundarySpan != nil {
-				r.splitMu.splitFinder.Record(*boundarySpan, rand.Intn)
+		shouldInitSplit := r.loadBasedSplitter.Record(timeutil.Now(), len(ba.Requests), func() roachpb.Span {
+			boundarySpan := spans.BoundarySpan(spanset.SpanGlobal)
+			if boundarySpan == nil {
+				return roachpb.Span{}
 			}
-		}
-		r.splitMu.Unlock()
-		// Add to the split queue after releasing splitMu.
-		if split {
+			return *boundarySpan
+		})
+		if shouldInitSplit {
 			r.store.splitQueue.MaybeAdd(r, r.store.Clock().Now())
 		}
 	}

--- a/pkg/storage/replica_init.go
+++ b/pkg/storage/replica_init.go
@@ -16,6 +16,7 @@ package storage
 
 import (
 	"context"
+	"math/rand"
 	"os"
 	"time"
 
@@ -23,6 +24,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/storage/abortspan"
 	"github.com/cockroachdb/cockroach/pkg/storage/spanlatch"
+	"github.com/cockroachdb/cockroach/pkg/storage/split"
 	"github.com/cockroachdb/cockroach/pkg/storage/stateloader"
 	"github.com/cockroachdb/cockroach/pkg/storage/storagebase"
 	"github.com/cockroachdb/cockroach/pkg/storage/txnwait"
@@ -46,6 +48,9 @@ func newReplica(rangeID roachpb.RangeID, store *Store) *Replica {
 	r.mu.stateLoader = stateloader.Make(rangeID)
 	r.mu.quiescent = true
 	r.mu.zone = config.DefaultZoneConfigRef()
+	split.Init(&r.loadBasedSplitter, rand.Intn, func() float64 {
+		return float64(SplitByLoadQPSThreshold.Get(&store.cfg.Settings.SV))
+	})
 
 	if leaseHistoryMaxEntries > 0 {
 		r.leaseHistory = newLeaseHistory()

--- a/pkg/storage/replica_split_load.go
+++ b/pkg/storage/replica_split_load.go
@@ -15,11 +15,8 @@
 package storage
 
 import (
-	"time"
-
 	"github.com/cockroachdb/cockroach/pkg/settings"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
-	"github.com/cockroachdb/cockroach/pkg/storage/split"
 )
 
 // SplitByLoadEnabled wraps "kv.range_split.by_load_enabled".
@@ -48,42 +45,4 @@ func (r *Replica) SplitByLoadEnabled() bool {
 	return SplitByLoadEnabled.Get(&r.store.cfg.Settings.SV) &&
 		r.store.ClusterSettings().Version.IsActive(cluster.VersionLoadSplits) &&
 		!r.store.TestingKnobs().DisableLoadBasedSplitting
-}
-
-// needsSplitByLoadLocked returns two bools indicating first, whether
-// the range is over the threshold for splitting by load and second,
-// whether the range is ready to be added to the split queue.
-func (r *Replica) needsSplitByLoadLocked() (bool, bool) {
-	// First compute requests per second since the last check.
-	nowTime := r.store.Clock().PhysicalTime()
-	duration := nowTime.Sub(r.splitMu.lastReqTime)
-	if duration < time.Second {
-		return r.splitMu.splitFinder != nil, false
-	}
-
-	// Update the QPS and reset the time and request counter.
-	r.splitMu.qps = (float64(r.splitMu.count) / float64(duration)) * 1e9
-	r.splitMu.lastReqTime = nowTime
-	r.splitMu.count = 0
-
-	// If the QPS for the range exceeds the threshold, start actively
-	// tracking potential for splitting this range based on load.
-	// This tracking will begin by initiating a splitFinder so it can
-	// begin to Record requests so it can find a split point. If a
-	// splitFinder already exists, we check if a split point is ready
-	// to be used.
-	if r.splitMu.qps >= r.SplitByLoadQPSThreshold() {
-		if r.splitMu.splitFinder != nil {
-			if r.splitMu.splitFinder.Ready(nowTime) {
-				// We're ready to add this range to the split queue.
-				return true, true
-			}
-		} else {
-			r.splitMu.splitFinder = split.New(nowTime)
-		}
-		return true, false
-	}
-
-	r.splitMu.splitFinder = nil
-	return false, false
 }

--- a/pkg/storage/split/decider.go
+++ b/pkg/storage/split/decider.go
@@ -1,0 +1,151 @@
+// Copyright 2019 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// TODO(tbg): rename this package. `lbsplit`?
+
+package split
+
+import (
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+)
+
+const minSplitSuggestionInterval = time.Minute
+
+// A Decider collects measurements about the activity (measured in qps) on a
+// Replica and, assuming that qps thresholds are exceeded, tries to determine
+// a split key that would approximately result in halving the load on each of
+// the resultant ranges.
+//
+// Operations should call `Record` with a current timestamp. Operation counts
+// are aggregated over a second and a qps computed. If the QPS is above threshold,
+// a split finder is instantiated and the spans supplied to Record are sampled
+// for a duration (on the order of ten seconds). Assuming that load consistently
+// remains over threshold, and the workload touches a diverse enough set of keys
+// to benefit from a split, sampling will eventually instruct a caller of Record
+// to carry out a split. When the split is initiated, it can obtain the suggested
+// split point from MaybeSplitKey (which may have disappeared either due to a drop
+// in qps or a change in the workload).
+type Decider struct {
+	intn         func(n int) int // supplied to Init
+	qpsThreshold func() float64  // supplied to Init
+
+	mu struct {
+		syncutil.Mutex
+		lastQPSRollover time.Time // most recent time recorded by requests.
+		qps             float64   // last reqs/s rate as of lastQPSRollover
+
+		count               int64     // number of requests recorded since last rollover
+		splitFinder         *Finder   // populated when engaged or decided
+		lastSplitSuggestion time.Time // last stipulation to client to carry out split
+	}
+}
+
+// Init initializes a Decider (which is assumed to be zero). The signature allows
+// embedding the Decider into a larger struct outside of the scope of this package
+// without incurring a pointer reference. This is relevant since many Deciders
+// may exist in the system at any given point in time.
+func Init(lbs *Decider, intn func(n int) int, qpsThreshold func() float64) {
+	lbs.intn = intn
+	lbs.qpsThreshold = qpsThreshold
+}
+
+// Record notifies the Decider that 'n' operations are being carried out which
+// operate on the span returned by the supplied method. The closure will only
+// be called when necessary, that is, when the Decider is considering a split
+// and is sampling key spans to determine a suitable split point.
+//
+// If the returned boolean is true, a split key is available (though it may
+// disappear as more keys are sampled) and should be initiated by the caller,
+// which can call MaybeSplitKey to retrieve the suggested key.
+func (d *Decider) Record(now time.Time, n int, span func() roachpb.Span) bool {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	return d.recordLocked(now, n, span)
+}
+
+func (d *Decider) recordLocked(now time.Time, n int, span func() roachpb.Span) bool {
+	d.mu.count += int64(n)
+
+	// First compute requests per second since the last check.
+	elapsedSinceLastQPS := now.Sub(d.mu.lastQPSRollover)
+	if elapsedSinceLastQPS >= time.Second {
+		if elapsedSinceLastQPS > 2*time.Second {
+			// Force a QPS of zero; there wasn't any activity within the last
+			// second at all.
+			d.mu.count = 0
+		}
+		// Update the QPS and reset the time and request counter.
+		d.mu.qps = (float64(d.mu.count) / float64(elapsedSinceLastQPS)) * 1e9
+		d.mu.lastQPSRollover = now
+		d.mu.count = 0
+
+		// If the QPS for the range exceeds the threshold, start actively
+		// tracking potential for splitting this range based on load.
+		// This tracking will begin by initiating a splitFinder so it can
+		// begin to Record requests so it can find a split point. If a
+		// splitFinder already exists, we check if a split point is ready
+		// to be used.
+		if d.mu.qps >= d.qpsThreshold() {
+			if d.mu.splitFinder == nil {
+				d.mu.splitFinder = NewFinder(now)
+			}
+		} else {
+			d.mu.splitFinder = nil
+		}
+	}
+
+	if d.mu.splitFinder != nil && n != 0 {
+		s := span()
+		if s.Key != nil {
+			d.mu.splitFinder.Record(span(), d.intn)
+		}
+		if now.Sub(d.mu.lastSplitSuggestion) > minSplitSuggestionInterval && d.mu.splitFinder.Ready(now) && d.mu.splitFinder.Key() != nil {
+			d.mu.lastSplitSuggestion = now
+			return true
+		}
+	}
+	return false
+}
+
+// LastQPS returns the most recent QPS measurement.
+func (d *Decider) LastQPS(now time.Time) float64 {
+	d.mu.Lock()
+	d.recordLocked(now, 0, nil)
+	qps := d.mu.qps
+	d.mu.Unlock()
+
+	return qps
+}
+
+// MaybeSplitKey returns a key to perform a split at. The return value will be
+// nil if either the Decider hasn't decided that a split should be carried out
+// or if it wasn't able to determine a suitable split key.
+//
+// It is legal to call MaybeSplitKey at any time.
+func (d *Decider) MaybeSplitKey(now time.Time) roachpb.Key {
+	var key roachpb.Key
+
+	d.mu.Lock()
+	d.recordLocked(now, 0, nil)
+	if d.mu.splitFinder != nil && d.mu.splitFinder.Ready(now) {
+		key = d.mu.splitFinder.Key()
+	}
+	d.mu.Unlock()
+
+	return key
+}

--- a/pkg/storage/split/decider_test.go
+++ b/pkg/storage/split/decider_test.go
@@ -1,0 +1,158 @@
+// Copyright 2019 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package split
+
+import (
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestDecider(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	intn := rand.New(rand.NewSource(12)).Intn
+
+	var d Decider
+	Init(&d, intn, func() float64 { return 10.0 })
+
+	ms := func(i int) time.Time {
+		ts, err := time.Parse(time.RFC3339, "2000-01-01T00:00:00Z")
+		assert.NoError(t, err)
+		return ts.Add(time.Duration(i) * time.Millisecond)
+	}
+
+	op := func(s string) func() roachpb.Span {
+		return func() roachpb.Span { return roachpb.Span{Key: roachpb.Key(s)} }
+	}
+
+	assertQPS := func(i int, expQPS float64) {
+		t.Helper()
+		qps := d.LastQPS(ms(i))
+		assert.Equal(t, expQPS, qps)
+	}
+
+	assert.Equal(t, false, d.Record(ms(100), 1, nil))
+	assertQPS(100, 0)
+
+	// The first operation was interpreted as having happened after an eternity
+	// of no activity, and rolled over the qps to mark the beginning of a new
+	// second. The next qps computation is expected for timestamps >= 1100.
+	assert.Equal(t, ms(100), d.mu.lastQPSRollover)
+	assert.EqualValues(t, 0, d.mu.count)
+
+	assert.Equal(t, false, d.Record(ms(400), 4, nil))
+	assertQPS(100, 0)
+	assertQPS(700, 0)
+
+	assert.Equal(t, false, d.Record(ms(300), 3, nil))
+	assertQPS(100, 0)
+
+	assert.Equal(t, false, d.Record(ms(900), 1, nil))
+	assertQPS(0, 0)
+
+	assert.Equal(t, false, d.Record(ms(1099), 1, nil))
+	assertQPS(0, 0)
+
+	// Now 9 operations happened in the interval [100, 1099]. The next higher
+	// timestamp will decide whether to engage the split finder.
+
+	// It won't engage because the duration between the rollovers is 1.1s, and
+	// we had 10 events over that interval.
+	assert.Equal(t, false, d.Record(ms(1200), 1, nil))
+	assertQPS(0, float64(10)/float64(1.1))
+	assert.Equal(t, ms(1200), d.mu.lastQPSRollover)
+
+	var nilFinder *Finder
+
+	assert.Equal(t, nilFinder, d.mu.splitFinder)
+
+	assert.Equal(t, false, d.Record(ms(2199), 12, nil))
+	assert.Equal(t, nilFinder, d.mu.splitFinder)
+
+	// 2200 is the next rollover point, and 12+1=13 qps should be computed.
+	assert.Equal(t, false, d.Record(ms(2200), 1, op("a")))
+	assert.Equal(t, ms(2200), d.mu.lastQPSRollover)
+	assertQPS(0, float64(13))
+
+	assert.NotNil(t, d.mu.splitFinder)
+	assert.False(t, d.mu.splitFinder.Ready(ms(10)))
+
+	// With continued partitioned write load, split finder eventually tells us
+	// to split. We don't test the details of exactly when that happens because
+	// this is done in the finder tests.
+	tick := 2200
+	for o := op("a"); !d.Record(ms(tick), 11, o); tick += 1000 {
+		if tick/1000%2 == 0 {
+			o = op("z")
+		} else {
+			o = op("a")
+		}
+	}
+
+	assert.Equal(t, roachpb.Key("z"), d.MaybeSplitKey(ms(tick)))
+
+	// We were told to split, but won't be told to split again for some time
+	// to avoid busy-looping on split attempts.
+	for i := 0; i <= int(minSplitSuggestionInterval/time.Second); i++ {
+		o := op("z")
+		if i%2 != 0 {
+			o = op("a")
+		}
+		assert.False(t, d.Record(ms(tick), 11, o))
+		assert.True(t, d.LastQPS(ms(tick)) > 1.0)
+		// Even though the split key remains.
+		assert.Equal(t, roachpb.Key("z"), d.MaybeSplitKey(ms(tick+999)))
+		tick += 1000
+	}
+	// But after minSplitSuggestionInterval of ticks, we get another one.
+	assert.True(t, d.Record(ms(tick), 11, op("a")))
+	assert.True(t, d.LastQPS(ms(tick)) > 1.0)
+
+	// Split key suggestion vanishes once qps drops.
+	tick += 1000
+	assert.False(t, d.Record(ms(tick), 9, op("a")))
+	assert.Equal(t, roachpb.Key(nil), d.MaybeSplitKey(ms(tick)))
+	assert.Equal(t, nilFinder, d.mu.splitFinder)
+
+	// Hammer a key with writes above threshold. There shouldn't be a split
+	// since everyone is hitting the same key and load can't be balanced.
+	for i := 0; i < 1000; i++ {
+		assert.False(t, d.Record(ms(tick), 11, op("q")))
+		tick += 1000
+	}
+	assert.True(t, d.mu.splitFinder.Ready(ms(tick)))
+	assert.Equal(t, roachpb.Key(nil), d.MaybeSplitKey(ms(tick)))
+
+	// But the finder keeps sampling to adapt to changing workload...
+	for i := 0; i < 1000; i++ {
+		assert.False(t, d.Record(ms(tick), 11, op("p")))
+		tick += 1000
+	}
+
+	// ... which we verify by looking at its samples directly.
+	for _, sample := range d.mu.splitFinder.samples {
+		assert.Equal(t, roachpb.Key("p"), sample.key)
+	}
+
+	// Since the new workload is also not partitionable, nothing changes in
+	// the decision.
+	assert.True(t, d.mu.splitFinder.Ready(ms(tick)))
+	assert.Equal(t, roachpb.Key(nil), d.MaybeSplitKey(ms(tick)))
+}

--- a/pkg/storage/split/finder.go
+++ b/pkg/storage/split/finder.go
@@ -69,8 +69,8 @@ type Finder struct {
 	count     int
 }
 
-// New initiates a Finder with the given time.
-func New(startTime time.Time) *Finder {
+// NewFinder initiates a Finder with the given time.
+func NewFinder(startTime time.Time) *Finder {
 	return &Finder{
 		startTime: startTime,
 	}
@@ -122,11 +122,11 @@ func (f *Finder) Record(span roachpb.Span, intNFn func(int) int) {
 	f.samples[idx] = sample{key: span.Key}
 }
 
-// Key finds an appropriate split point based on the Reservoir
-// sampling method.
-func (f *Finder) Key() (bool, roachpb.Key) {
+// Key finds an appropriate split point based on the Reservoir sampling method.
+// Returns a nil key if no appropriate key was found.
+func (f *Finder) Key() roachpb.Key {
 	if f == nil {
-		return false, nil
+		return nil
 	}
 
 	var bestIdx = -1
@@ -149,7 +149,7 @@ func (f *Finder) Key() (bool, roachpb.Key) {
 	}
 
 	if bestIdx == -1 {
-		return false, nil
+		return nil
 	}
-	return true, f.samples[bestIdx].key
+	return f.samples[bestIdx].key
 }

--- a/pkg/storage/split/finder_test.go
+++ b/pkg/storage/split/finder_test.go
@@ -133,35 +133,33 @@ func TestSplitFinderKey(t *testing.T) {
 
 	testCases := []struct {
 		reservoir      [splitKeySampleSize]sample
-		splitByLoad    bool
 		splitByLoadKey roachpb.Key
 	}{
 		// Test an empty reservoir.
-		{basicReservoir, false, nil},
+		{basicReservoir, nil},
 		// Test reservoir with no load should have no splits.
-		{noLoadReservoir, false, nil},
+		{noLoadReservoir, nil},
 		// Test a uniform reservoir (Splits at the first key)
-		{uniformReservoir, true, keys.MakeTablePrefix(ReservoirKeyOffset)},
+		{uniformReservoir, keys.MakeTablePrefix(ReservoirKeyOffset)},
 		// Testing a non-uniform reservoir.
-		{nonUniformReservoir, true, keys.MakeTablePrefix(ReservoirKeyOffset + splitKeySampleSize/2)},
+		{nonUniformReservoir, keys.MakeTablePrefix(ReservoirKeyOffset + splitKeySampleSize/2)},
 		// Test a load heavy reservoir on a single hot key. Splitting can't help here.
-		{singleHotKeyReservoir, false, nil},
+		{singleHotKeyReservoir, nil},
 		// Test a load heavy reservoir on multiple hot keys. Splits between the hot keys.
-		{multipleHotKeysReservoir, true, keys.MakeTablePrefix(ReservoirKeyOffset + 1)},
+		{multipleHotKeysReservoir, keys.MakeTablePrefix(ReservoirKeyOffset + 1)},
 		// Test a spanning reservoir. Splitting will be bad here. Should avoid it.
-		{spanningReservoir, false, nil},
+		{spanningReservoir, nil},
 		// Test that splits happen between two heavy spans.
-		{multipleSpanReservoir, true, keys.MakeTablePrefix(ReservoirKeyOffset + splitKeySampleSize/2)},
+		{multipleSpanReservoir, keys.MakeTablePrefix(ReservoirKeyOffset + splitKeySampleSize/2)},
 	}
 
 	for i, test := range testCases {
-		finder := New(timeutil.Now())
+		finder := NewFinder(timeutil.Now())
 		finder.samples = test.reservoir
-		if splitByLoad, splitByLoadKey := finder.Key(); splitByLoad != test.splitByLoad ||
-			!bytes.Equal(splitByLoadKey, test.splitByLoadKey) {
+		if splitByLoadKey := finder.Key(); !bytes.Equal(splitByLoadKey, test.splitByLoadKey) {
 			t.Errorf(
-				"%d: expected splitByLoad: %t and splitByLoadKey: %v, but got splitByLoad: %t and splitByLoadKey: %v",
-				i, test.splitByLoad, test.splitByLoadKey, splitByLoad, splitByLoadKey)
+				"%d: expected splitByLoadKey: %v, but got splitByLoadKey: %v",
+				i, test.splitByLoadKey, splitByLoadKey)
 		}
 	}
 }
@@ -266,7 +264,7 @@ func TestSplitFinderRecorder(t *testing.T) {
 	}
 
 	for i, test := range testCases {
-		finder := New(timeutil.Now())
+		finder := NewFinder(timeutil.Now())
 		finder.samples = test.currReservoir
 		finder.count = test.currCount
 		finder.Record(test.recordSpan, test.intNFn)

--- a/pkg/storage/split_queue.go
+++ b/pkg/storage/split_queue.go
@@ -27,6 +27,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/humanizeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/pkg/errors"
 )
 
@@ -88,12 +89,7 @@ func newSplitQueue(store *Store, db *client.DB, gossip *gossip.Gossip) *splitQue
 }
 
 func shouldSplitRange(
-	desc *roachpb.RangeDescriptor,
-	ms enginepb.MVCCStats,
-	reqRate float64,
-	splitByLoadQPSThreshold float64,
-	maxBytes int64,
-	sysCfg *config.SystemConfig,
+	desc *roachpb.RangeDescriptor, ms enginepb.MVCCStats, maxBytes int64, sysCfg *config.SystemConfig,
 ) (shouldQ bool, priority float64) {
 	if sysCfg.NeedsSplit(desc.StartKey, desc.EndKey) {
 		// Set priority to 1 in the event the range is split by zone configs.
@@ -108,12 +104,7 @@ func shouldSplitRange(
 		shouldQ = true
 	}
 
-	// Check if the request rate is higher than the QPS threshold.
-	if reqRate >= splitByLoadQPSThreshold {
-		priority += reqRate / splitByLoadQPSThreshold
-		shouldQ = true
-	}
-	return
+	return shouldQ, priority
 }
 
 // shouldQueue determines whether a range should be queued for
@@ -124,15 +115,11 @@ func (sq *splitQueue) shouldQueue(
 	ctx context.Context, now hlc.Timestamp, repl *Replica, sysCfg *config.SystemConfig,
 ) (shouldQ bool, priority float64) {
 	shouldQ, priority = shouldSplitRange(repl.Desc(), repl.GetMVCCStats(),
-		0, /* Don't check for load based splitting yet. */
-		repl.SplitByLoadQPSThreshold(), repl.GetMaxBytes(), sysCfg)
+		repl.GetMaxBytes(), sysCfg)
 
 	if !shouldQ && repl.SplitByLoadEnabled() {
-		repl.splitMu.Lock()
-		defer repl.splitMu.Unlock()
-		if splitByLoad, _ := repl.splitMu.splitFinder.Key(); splitByLoad {
-			priority++
-			shouldQ = true
+		if splitKey := repl.loadBasedSplitter.MaybeSplitKey(timeutil.Now()); splitKey != nil {
+			shouldQ, priority = true, 1.0 // default priority
 		}
 	}
 
@@ -205,30 +192,11 @@ func (sq *splitQueue) processAttempt(
 		return err
 	}
 
-	// Decide whether to split by load. This is the case if there's an engaged
-	// split finder which is Ready() and has identified a key suitable for
-	// splitting at.
-	var splitByLoadKey roachpb.Key
-	var splitQPS float64
-
-	// If this replica was added to the split queue based on load, we want to
-	// make sure the criteria still hold. For example, if a range was added for
-	// a split due to a 10s write burst that suddenly stopped, we don't want to
-	// split it when it is processed a minute later.
-	r.splitMu.Lock()
-	if _, shouldSplit := r.needsSplitByLoadLocked(); shouldSplit {
-		var splitByLoad bool
-		splitByLoad, splitByLoadKey = r.splitMu.splitFinder.Key()
-		splitQPS = r.splitMu.qps
-		if splitByLoad {
-			r.splitMu.splitFinder = nil
-		}
-	}
-	r.splitMu.Unlock()
-
-	if splitByLoadKey != nil {
+	now := timeutil.Now()
+	if splitByLoadKey := r.loadBasedSplitter.MaybeSplitKey(now); splitByLoadKey != nil {
 		batchHandledQPS := r.QueriesPerSecond()
 		raftAppliedQPS := r.WritesPerSecond()
+		splitQPS := r.loadBasedSplitter.LastQPS(now)
 		reason := fmt.Sprintf(
 			"load at key %s (%.2f splitQPS, %.2f batches/sec, %.2f raft mutations/sec)",
 			splitByLoadKey,

--- a/pkg/storage/split_queue_test.go
+++ b/pkg/storage/split_queue_test.go
@@ -41,36 +41,31 @@ func TestSplitQueueShouldQueue(t *testing.T) {
 	config.TestingSetZoneConfig(2000, config.ZoneConfig{RangeMaxBytes: proto.Int64(32 << 20)})
 	config.TestingSetZoneConfig(2002, config.ZoneConfig{RangeMaxBytes: proto.Int64(32 << 20)})
 
-	minQPSForSplit := SplitByLoadQPSThreshold.Get(&tc.store.ClusterSettings().SV)
-
 	testCases := []struct {
 		start, end roachpb.RKey
 		bytes      int64
-		QPS        float64
 		maxBytes   int64
 		shouldQ    bool
 		priority   float64
 	}{
 		// No intersection, no bytes, no load.
-		{roachpb.RKeyMin, roachpb.RKey(keys.MetaMax), 0, 0, 64 << 20, false, 0},
+		{roachpb.RKeyMin, roachpb.RKey(keys.MetaMax), 0, 64 << 20, false, 0},
 		// Intersection in zone, no bytes, no load.
-		{keys.MakeTablePrefix(2001), roachpb.RKeyMax, 0, 0, 64 << 20, true, 1},
+		{keys.MakeTablePrefix(2001), roachpb.RKeyMax, 0, 64 << 20, true, 1},
 		// Already split at largest ID, no load.
-		{keys.MakeTablePrefix(2002), roachpb.RKeyMax, 0, 0, 32 << 20, false, 0},
+		{keys.MakeTablePrefix(2002), roachpb.RKeyMax, 0, 32 << 20, false, 0},
 		// Multiple intersections, no bytes, no load.
-		{roachpb.RKeyMin, roachpb.RKeyMax, 0, 0, 64 << 20, true, 1},
+		{roachpb.RKeyMin, roachpb.RKeyMax, 0, 64 << 20, true, 1},
 		// No intersection, max bytes, no load.
-		{roachpb.RKeyMin, roachpb.RKey(keys.MetaMax), 64 << 20, 0, 64 << 20, false, 0},
+		{roachpb.RKeyMin, roachpb.RKey(keys.MetaMax), 64 << 20, 64 << 20, false, 0},
 		// No intersection, max bytes+1, no load.
-		{roachpb.RKeyMin, roachpb.RKey(keys.MetaMax), 64<<20 + 1, 0, 64 << 20, true, 1},
+		{roachpb.RKeyMin, roachpb.RKey(keys.MetaMax), 64<<20 + 1, 64 << 20, true, 1},
 		// No intersection, max bytes * 2, no load.
-		{roachpb.RKeyMin, roachpb.RKey(keys.MetaMax), 64 << 21, 0, 64 << 20, true, 2},
+		{roachpb.RKeyMin, roachpb.RKey(keys.MetaMax), 64 << 21, 64 << 20, true, 2},
 		// Intersection, max bytes +1, no load.
-		{keys.MakeTablePrefix(2000), roachpb.RKeyMax, 32<<20 + 1, 0, 32 << 20, true, 2},
+		{keys.MakeTablePrefix(2000), roachpb.RKeyMax, 32<<20 + 1, 32 << 20, true, 2},
 		// Split needed at table boundary, but no zone config, no load.
-		{keys.MakeTablePrefix(2001), roachpb.RKeyMax, 32<<20 + 1, 0, 64 << 20, true, 1},
-		// Split needed because of load.
-		{roachpb.RKeyMin, roachpb.RKey(keys.MetaMax), 0, float64(minQPSForSplit), 64 << 20, true, 1},
+		{keys.MakeTablePrefix(2001), roachpb.RKeyMax, 32<<20 + 1, 64 << 20, true, 1},
 	}
 
 	cfg := tc.gossip.GetSystemConfig()
@@ -93,9 +88,6 @@ func TestSplitQueueShouldQueue(t *testing.T) {
 		repl.mu.Lock()
 		repl.mu.state.Stats = &enginepb.MVCCStats{KeyBytes: test.bytes}
 		repl.mu.Unlock()
-		repl.splitMu.Lock()
-		repl.splitMu.qps = test.QPS
-		repl.splitMu.Unlock()
 		zoneConfig := config.DefaultZoneConfig()
 		zoneConfig.RangeMaxBytes = proto.Int64(test.maxBytes)
 		repl.SetZoneConfig(&zoneConfig)
@@ -103,8 +95,7 @@ func TestSplitQueueShouldQueue(t *testing.T) {
 		// Testing using shouldSplitRange instead of shouldQueue to avoid using the splitFinder
 		// This tests the merge queue behavior too as a result. For splitFinder tests,
 		// see split/split_test.go.
-		shouldQ, priority := shouldSplitRange(repl.Desc(), repl.GetMVCCStats(),
-			repl.GetSplitQPS(), repl.SplitByLoadQPSThreshold(), repl.GetMaxBytes(), cfg)
+		shouldQ, priority := shouldSplitRange(repl.Desc(), repl.GetMVCCStats(), repl.GetMaxBytes(), cfg)
 		if shouldQ != test.shouldQ {
 			t.Errorf("%d: should queue expected %t; got %t", i, test.shouldQ, shouldQ)
 		}
@@ -113,8 +104,3 @@ func TestSplitQueueShouldQueue(t *testing.T) {
 		}
 	}
 }
-
-////
-// NOTE: tests which actually verify processing of the split queue are
-// in client_split_test.go, which is in a different test package in
-// order to allow for distributed transactions with a proper client.

